### PR TITLE
New test feedback generation

### PR DESCRIPTION
**Pull request by densh**
Previously we used to generate test feedback immediately during the test run. Unfortunately this didn't quite work in situations where a fatal jvm crash happend (e.g. StackOverflow) as test file was written only on completion of the entire suite. 

This pr introduces new two-phase grading feedback generation:
1. Serialize a log of relevant events into a file while running scalatest tests. 
2. In a separate process read this file, replay events and generate a grading summary.

This makes it proof against JVM crashes during testing phase. 

Review @vjovanov.

### CODE REVIEW

File: runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        "Your solution achieved a testing score of %d out of %d.\n\n".format(score, maxScore)
+      }
+      if (score == maxScore)
+        sb.append("Great job!!!\n\n")
+      else {
+        sb.append("""Below you can see a short feedback for every test that failed,
+                    |indicating the reason for the test failure and how many points
+                    |you lost for each individual test.
+                    |
+                    |Tests that were aborted took too long too complete or crashed the
```
Comment (vjovanov) [inline runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala:54]: Can we improve the formulation here? I think we are missing an action here. It should be "tests that x y z are ?."

File: runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+      case TestSuccess(TestId(suiteName, name, _)) =>
+        suites(suiteName).tests(name).failure = None
+      case TestFailure(TestId(suiteName, name, _), msg: String) =>
+        suites(suiteName).tests(name).failure = Some(msg)
+    }
+    GradingSummary(suites.values.toList)
+  }
+}
+
+object GradingSummaryRunner {
```
Comment (vjovanov) [inline runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala]: Why do we need this? For testing?

File: runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  final case class Suite(val name: String, val weight: Int,
+    var complete: Boolean = false,
+    val tests: mutable.Map[String, Test] = mutable.Map.empty)
+
+  final case class Test(val name: String, val weight: Int,
+    var failure: Option[String])
+
+  /** Given test completion state, compute the grade and provide feedback. */
+  def apply(suites: List[Suite]): GradingSummary = {
+    val score = suites.map { _.tests.values.map { t => t.failure.fold(t.weight)(_ => 0) }.sum }.sum
```
Comment (vjovanov) [inline runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala]: Too long line.

File: runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    val maxScore = suites.map { _.weight }.sum
```
Comment (vjovanov) [inline runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala]: Why not `suites.map(_.weight).sum`?

**Comment by vjovanov** [comment]
From the code it is not obvious where you detect tests that started and never completed. I also do not see the custom message for these tests.

**Comment by vjovanov** [comment]
Otherwise LGTM.

File: runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    val entries = jsons.map(_.unpickle[Entry])
+    val suites = mutable.Map.empty[String, Suite]
+    entries.foreach {
+      case SuiteStart(SuiteId(name, ToInt(weight))) =>
+        suites += name -> Suite(name, weight)
+      case SuiteEnd(SuiteId(name, _)) =>
+        suites(name).complete = true
+      case TestStart(TestId(suiteName, name, ToInt(weight))) =>
+        suites(suiteName).tests +=
+          name -> Test(name, weight, failure = Some("test has been aborted"))
```
Comment (densh) [inline runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala:87]: @vjovanov By default tests have "test has been aborted" failure until matching TestSuccess or TestFailure happens. This means that there is no need for extra post-processing to weed out aborted tests.

File: runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala (hunk 4 above)
Comment (densh) [inline runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala]: Fixed.

File: runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala (hunk 3 above)
Comment (densh) [inline runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala]: Fixed.

File: runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala (hunk 1 above)
Comment (densh) [inline runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala:54]: It refers to "test has been aborted" failure. How would you reformulate the sentence? It's not quite clear to me.

**Comment by densh** [comment]
@vjovanov can you also have a look at https://github.com/sbt-coursera/sbt-coursera/pull/17 ?

File: runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala (hunk 1 above)
Comment (vjovanov) [inline runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala:54]: Aha, now it makes sense.

File: runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala (hunk 5 above)
Comment (vjovanov) [inline runtime/src/main/scala/ch/epfl/lamp/grading/GradingSummary.scala:87]: Great!

**Comment by vjovanov** [comment]
LGTM

File: runtime/src/main/scala/ch/epfl/lamp/grading/GradingReporter.scala [hunk 6]
```diff
@@ -1,156 +1,89 @@
...
+     * scalatest. We are not interested in it.
+     */
+    case e: SuiteStarting if e.suiteName != "DiscoverySuite" => record(SuiteStart(e.suiteName))
+    case e: SuiteCompleted if e.suiteName != "DiscoverySuite" => record(SuiteEnd(e.suiteName))
+    /* We don't get a `TestStarting` for ignored tests, but we do get here
+     * one tests that use `pending`.
+     */
+    case e: TestStarting => record(TestStart(e.testName))
+    case e: TestSucceeded => record(TestSuccess(e.testName))
+    case e: TestPending => record(TestSuccess(e.testName))
```
Comment (vjovanov) [inline runtime/src/main/scala/ch/epfl/lamp/grading/GradingReporter.scala:71]: @densh 

Is this correct?
Should it be TestPending?

**Pull request merged**